### PR TITLE
skip converting empty yaml objects

### PR DESCRIPTION
**Pull request by ashwgit**
PTAL, this one skips empty YAML documents during FBC YAML→JSON conversion to prevent emitting null objects that breaks opm validate while performing add operations.
Also fixes a bug for including conversion for .yml file too.

Opened For - https://github.com/release-engineering/iib/pull/1346

### CODE REVIEW

File: tests/test_workers/test_tasks/test_fbc_utils.py [hunk 1]
```diff
@@ -300,3 +300,53 @@ def test__serialize_datetime():
...
+    open(empty_yaml, 'w').close()
+
+    with pytest.raises(IIBError) as exc_info:
+        enforce_json_config_dir(str(tmpdir))
+
+    assert "Empty YAML file found" in str(exc_info.value)
+    assert "empty.yaml" in str(exc_info.value)
+
+
+def test_enforce_json_config_dir_handles_yml_extension(tmpdir):
```
[the remark this answers is not in the document]
Comment (lipoja) [reply tests/test_workers/test_tasks/test_fbc_utils.py:358]: @ashwgit Do you consider implementing this one?

File: iib/workers/tasks/fbc_utils.py [hunk 2]
```diff
@@ -169,5 +173,8 @@ def enforce_json_config_dir(config_dir: str) -> None:
                 with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
                     data = yaml.load_all(yaml_in)
                     for chunk in data:
```
[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/fbc_utils.py:176]: I think we should consider implementing this.

[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:176]: @lipoja  implemented , PTAL

File: tests/test_workers/test_tasks/test_fbc_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashwgit) [reply tests/test_workers/test_tasks/test_fbc_utils.py:358]: added

File: iib/workers/tasks/fbc_utils.py [hunk 3]
```diff
@@ -154,12 +154,16 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
     """
     log.info("Enforcing JSON content on config_dir: %s", config_dir)
     for dirpath, _, filenames in os.walk(config_dir):
         for file in filenames:
             in_file = os.path.join(dirpath, file)
-            if in_file.lower().endswith(".yaml"):
+            if in_file.lower().endswith(".yaml") or in_file.lower().endswith(".yml"):
+                if os.path.getsize(in_file) == 0:
+                    raise IIBError(f"Empty YAML file found: {in_file}")
+
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:166]: I don't think we would ever be having two files with name name but different extensions. @chandwanitulsi WDYT ?

[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/fbc_utils.py:178]: In theory we can have multiple files inside <package-name> directory, it can contain also  directory structure with yaml/json files. The thing is those files should not contain duplicate data. So no duplicate package could be found. 

The silent agreement is that IIB forces to have one rendered catalog file per package. 

That means it can happen that fbc-fragment contains multiple files. I am not sure if there is any enforcement during the pipeline run to have only one file per package. We were working on something but I cant remember if we enforced that or modified IIB to pass.

[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:178]: so if we agree that we won't have two files with same name and different extension of "yaml" and "yml" then we can ignore this part, I don't want to increase cold paths for code and it will effect maintainability of code.

File: iib/workers/tasks/fbc_utils.py [hunk 4]
```diff
@@ -154,20 +154,34 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
                 out_file = os.path.join(dirpath, f"{Path(in_file).stem}.json")
                 log.debug(f"Converting {in_file} to {out_file}.")
                 # Make sure the output file doesn't exist before opening in append mode
                 with contextlib.suppress(FileNotFoundError):
                     os.remove(out_file)
                 # The input file may contain multiple chunks, we must append them accordingly
+                wrote_chunk = False
                 with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
                     data = yaml.load_all(yaml_in)
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py]: I would rather create a new MR for this one instead handling everything at once at it decrease the delivery speed. @chandwanitulsi wdyt ?

File: iib/workers/tasks/fbc_utils.py [hunk 5]
```diff
@@ -154,20 +154,34 @@ def enforce_json_config_dir(config_dir: str) -> None:
     It will walk recursively and convert any YAML files to the JSON format.
 
     :param str config_dir: The config dir to walk recursively converting any YAML to JSON.
+    :raises IIBError: If the yaml content is empty.
     """
     log.info("Enforcing JSON content on config_dir: %s", config_dir)
     for dirpath, _, filenames in os.walk(config_dir):
         for file in filenames:
             in_file = os.path.join(dirpath, file)
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:174]: I will discuss the decoupling with team and if agreed we will split this function into multiple small func.

[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:174]: I will mark the build as fail and there won't be any changes to git directory.

File: iib/workers/tasks/fbc_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/fbc_utils.py]: Pre-existing issue - lets skip for now.

File: iib/workers/tasks/fbc_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lipoja) [reply iib/workers/tasks/fbc_utils.py:178]: I do not agree that this will not happen, I think it might happen however the chances are really low. 
If you don't want to implement it it is up to you. 

I think it would not hurt to implement this suggestion:

> Before conversion, detect if both `<stem>.yaml` and `<stem>.yml` exist in the same directory and raise `IIBError` with a clear message (preferred to avoid silent data loss),

But it is up to you, I would be ok to skip it. Since the chances are low. 

@chandwanitulsi @MichalZelenak Do you have any opinion on this?

[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply iib/workers/tasks/fbc_utils.py:178]: yeah we cam have multiple files in the the config of one operator, but I don't suppose we would have same file with both .yaml and .yml extensions, that sounds rare

File: tests/test_workers/test_tasks/test_fbc_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chandwanitulsi) [reply tests/test_workers/test_tasks/test_fbc_utils.py:358]: @ashwgit can you also add a test where there is one file yaml file, but different yamls in the same file and some of those are empty? that was the exact failure case when this bug was discovered

File: iib/workers/tasks/fbc_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:178]: @lipoja thanks, I have added it.

File: iib/workers/tasks/fbc_utils.py [hunk 6]
```diff
@@ -154,20 +154,46 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
                 out_file = os.path.join(dirpath, f"{Path(in_file).stem}.json")
                 log.debug(f"Converting {in_file} to {out_file}.")
                 # Make sure the output file doesn't exist before opening in append mode
                 with contextlib.suppress(FileNotFoundError):
                     os.remove(out_file)
                 # The input file may contain multiple chunks, we must append them accordingly
+                wrote_chunk = False
                 with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
                     data = yaml.load_all(yaml_in)
                     for chunk in data:
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py]: added.

File: iib/workers/tasks/fbc_utils.py [hunk 7]
```diff
@@ -154,20 +154,46 @@ def enforce_json_config_dir(config_dir: str) -> None:
     It will walk recursively and convert any YAML files to the JSON format.
 
     :param str config_dir: The config dir to walk recursively converting any YAML to JSON.
+    :raises IIBError: If the yaml content is empty.
     """
     log.info("Enforcing JSON content on config_dir: %s", config_dir)
     for dirpath, _, filenames in os.walk(config_dir):
+        yaml_stems: dict[str, str] = {}
+        for file in filenames:
+            lower = file.lower()
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:163]: It's a expected implementation, this should be fine.

File: iib/workers/tasks/fbc_utils.py [hunk 8]
```diff
@@ -154,20 +154,50 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
+                    with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
+                        data = yaml.load_all(yaml_in)
+                        for chunk in data:
+                            # Ignore if it is an empty yaml object
+                            if chunk is None:
+                                continue
+                            json.dump(chunk, json_out, default=_serialize_datetime)
+                            wrote_chunk = True
+                except ruamel.yaml.YAMLError as e:
+                    with contextlib.suppress(FileNotFoundError):
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:196]: low, won't be implemented.

File: iib/workers/tasks/fbc_utils.py [hunk 9]
```diff
@@ -154,20 +154,50 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
     :param str config_dir: The config dir to walk recursively converting any YAML to JSON.
+    :raises IIBError: If the yaml content is empty.
     """
     log.info("Enforcing JSON content on config_dir: %s", config_dir)
     for dirpath, _, filenames in os.walk(config_dir):
+        yaml_stems: dict[str, str] = {}
+        for file in filenames:
+            lower = file.lower()
+            if lower.endswith(".yaml") or lower.endswith(".yml"):
+                stem = Path(file).stem
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:165]: Not required.

File: iib/workers/tasks/fbc_utils.py [hunk 10]
```diff
@@ -154,20 +154,50 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
+                        f"{yaml_stems[stem]} and {file} would both produce {stem}.json"
+                    )
+                yaml_stems[stem] = file
+
         for file in filenames:
             in_file = os.path.join(dirpath, file)
-            if in_file.lower().endswith(".yaml"):
+            if in_file.lower().endswith(".yaml") or in_file.lower().endswith(".yml"):
+                if os.path.getsize(in_file) == 0:
+                    raise IIBError(f"Empty YAML file found: {in_file}")
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:177]: Not required.

File: iib/workers/tasks/fbc_utils.py [hunk 11]
```diff
@@ -154,20 +154,50 @@ def enforce_json_config_dir(config_dir: str) -> None:
     It will walk recursively and convert any YAML files to the JSON format.
 
     :param str config_dir: The config dir to walk recursively converting any YAML to JSON.
+    :raises IIBError: If the yaml content is empty.
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:157]: Not required.

**Comment by ashwgit** [comment]
@lipoja @chandwanitulsi @MichalZelenak Please have a look again.

**Review by lipoja — APPROVED** [review]
